calyx/src/passes/compile_sync.rs

```diff
@@ -0,0 +1,380 @@
+    let mut assigns = build_assignments!(builder;
+        // barrier_*.write_en_0 = 1'd1;
+        barrier["write_en_0"] = ?cst_1["out"];
+        // barrier_*.in_0 = 32'd0;
+        barrier["in_0"] = ?cst_2["out"];
+        // restore_*[done] = barrier_*.write_done_0;
+        group["done"] = ?barrier["write_done_0"];
+    );
+    group.borrow_mut().assignments.append(&mut assigns);
+}
+
+fn build_wait_restore(
+    builder: &mut ir::Builder,
+    group: &RRC<ir::Group>,
+    eq: &RRC<ir::Cell>,
+) {
+    structure!(builder;
+    let wait_restore_reg = prim std_reg(1);
+    let cst_1 = constant(1, 1););
+    let eq_guard = !guard!(eq["out"]);
+    let mut assigns = build_assignments!(builder;
+    // wait_restore_reg_*.in = !eq_*.out? 1'd1;
+    wait_restore_reg["in"] = eq_guard? cst_1["out"];
+    // wait_restore_reg_*.write_en = !eq_*.out? 1'd1;
+    wait_restore_reg["write_en"] = eq_guard? cst_1["out"];
+    //wait_restore_*[done] = wait_restore_reg_*.done;
+    group["done"] = ?wait_restore_reg["done"];
+    );
+    group.borrow_mut().assignments.append(&mut assigns);
+}
+
+fn build_member_0(
+    original: &ir::Control,
+    incr_barrier: RRC<ir::Group>,
+    write_barrier: RRC<ir::Group>,
+    wait: RRC<ir::Group>,
+    clear_barrier: RRC<ir::Group>,
+    restore: RRC<ir::Group>,
+) -> ir::Control {
+    let mut stmts: Vec<ir::Control> = Vec::new();
+    let mut copy = ir::Control::clone(original);
+
+    copy.get_mut_attributes().unwrap().remove("sync");
+
+    stmts.push(copy);
+    stmts.push(ir::Control::enable(incr_barrier));
+    stmts.push(ir::Control::enable(write_barrier));
+    stmts.push(ir::Control::enable(wait));
+    stmts.push(ir::Control::enable(clear_barrier));
+    stmts.push(ir::Control::enable(restore));
+    ir::Control::seq(stmts)
+}
+
+fn build_member(
+    original: &ir::Control,
+    incr_barrier: RRC<ir::Group>,
+    write_barrier: RRC<ir::Group>,
+    wait: RRC<ir::Group>,
+    wait_restore: RRC<ir::Group>,
+) -> ir::Control {
+    let mut stmts: Vec<ir::Control> = Vec::new();
+    let mut copy = ir::Control::clone(original);
+
+    copy.get_mut_attributes().unwrap().remove("sync");
+
+    stmts.push(copy);
+    stmts.push(ir::Control::enable(incr_barrier));
+    stmts.push(ir::Control::enable(write_barrier));
+    stmts.push(ir::Control::enable(wait));
+    stmts.push(ir::Control::enable(wait_restore));
+    ir::Control::seq(stmts)
+}
+
+fn add_shared_structure(
+    builder: &mut ir::Builder,
+    barrier_idx: &u64,
+    map: &mut BarrierMap,
+) {
+    structure!(builder;
+            let barrier = prim std_sync_reg(32);
+            let eq = prim std_eq(32);
+    );
+    let wait_restore = builder.add_group("wait_restore");
+    let restore = builder.add_group("restore");
+    let clear_barrier = builder.add_group("clear_barrier");
+    build_restore(builder, &restore, &barrier);
+    build_wait_restore(builder, &wait_restore, &eq);
+    build_clear_barrier(builder, &clear_barrier, &barrier);
```

Comment:
These methods should return the `RRC<ir::Group>`  instead of taking a reference to them